learn_evm/arithmetic-checks.md

```diff
@@ -0,0 +1,668 @@
+
+> It's worth noting that Solidity aims to make the new Yul pipeline the standard.
+> Certain operations (including arithmetic checks) are always included as Yul code, regardless of whether the code is compiled with the new pipeline using `--via-ir`.
+> This provides an opportunity to examine the Yul code and gain a better understanding of how arithmetic checks are executed in Solidity.
+> However, it's important to keep in mind that the final bytecode may differ slightly when compiler optimizations are turned on.
+
+To illustrate how the compiler detects overflow in unsigned integer addition, consider the following example of Yul code that is produced by the compiler.
+
+```solidity
+function checked_add_t_uint256(x, y) -> sum {
+    x := cleanup_t_uint256(x)
+    y := cleanup_t_uint256(y)
+
+    // overflow, if x > (maxValue - y)
+    if gt(x, sub(0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff, y)) { panic_error_0x11() }
+
+    sum := add(x, y)
+}
+```
+
+To enhance readability, we can interpret the Yul code back into high-level Solidity code.
+
+```solidity
+/// @notice versions >=0.8.0 && <0.8.16
+function checkedAddUint1(uint256 a, uint256 b) public pure returns (uint256 c) {
+    unchecked {
+        c = a + b;
+
+        if (a > type(uint256).max - b) arithmeticError();
+    }
+}
+```
+
+> Solidity's arithmetic errors are encoded as `abi.encodeWithSignature("Panic(uint256)", 0x11)`.
+
+The check for overflow in unsigned integer addition involves calculating the largest value that one summand can be added to the other without resulting in an overflow.
+Specifically, in this case, the maximum value that `a` can have is `type(uint256).max - b`.
+If `a` exceeds this value, we can conclude that `a + b` will overflow.
+
+An alternative (and slightly more efficient) approach to computing the maximum value of `a` is to invert the bits on `b`.
+
+```solidity
+/// @notice versions >=0.8.0 && <0.8.16 with compiler optimizations
+function checkedAddUint2(uint256 a, uint256 b) public pure returns (uint256 c) {
+    unchecked {
+        c = a + b;
+
+        if (a > ~b) arithmeticError();
+    }
+}
+```
+
+This is equivalent, because `type(uint256).max` is a 256-bit integer with all its bits set to `1`.
+Subtracting `b` from `type(uint256).max` can be viewed as inverting each bit in `b`.
+This can be demonstrated by the transformation `~b = ~(0 ^ b) = ~0 ^ b = MAX ^ b = MAX - b`.
+
+> It's worth noting that `a - b = a ^ b` is **NOT** a general rule, except in special cases, such as when one of the values equals `MAX`.
+> We also obtain the relation `~b + 1 = 0 - b = -b` if we add `1` mod `2**256` to both sides of the previous equation.
+
+By computing the result of the addition first and then performing a check on the sum,
+modern versions of Solidity can eliminate the need for performing extra arithmetic operations in the comparison.
```

Comment:
- why `modern versions of Solidity`?  (0.8.16?)
- `can eliminate the need for performing extra arithmetic operations in the comparison` I am not sure what we mean here. My assumption is that its cheaper to do the `+` and then the check, that to do a separate check,  and then the `+`, but its not clear if its what we are trying to convey here.

Reply:
yes, 0.8.16 as seen in the code comment below. And that's pretty much what I mean. The code below doesn't need additional operations before the comparison. But I agree, it's a bit hard to understand, I'll try to re-phrase it.